sale_promotion_rule/models/sale_promotion_rule.py

```diff
@@ -0,0 +1,317 @@
+        self.remove_promotions(orders_without_coupon)
+        # coupon take precedence on auto rules
+        coupon_rule._apply(orders_without_coupon)
+        self.apply_auto(orders_without_coupon)
+
+    @api.model
+    def apply_auto(self, orders):
+        """Apply automatic promotion rules to the orders"""
+        auto_rules = self.search([
+            ('rule_type', '=', 'auto'),
+        ])
+        auto_rules._apply(orders)
+
+    @api.model
+    def remove_promotions(self, orders):
+        orders.write({
+            'promotion_rule_ids': [(5,)],
+            'coupon_promotion_rule_id': False
+        })
+        self._remove_promotions_lines(orders.mapped('order_line'))
+
+    @api.model
+    def _remove_promotions_lines(self, lines):
+        lines_by_order = defaultdict(self.env['sale.order.line'].browse)
+        for line in lines:
+            lines_by_order[line.order_id] |= line
+        # update lines from the order to avoid to trigger the compute
+        # methods on each line updated. Indeed, update on a X2many field
+        # is always done in norecompute on the parent...
+        for order, _lines in lines_by_order.items():
+            vals = []
+            for line in _lines:
+                if not line.has_promotion_rules:
+                    continue
+                v = {
+                    'discount': 0.0,
+                    'coupon_promotion_rule_id': False,
+                    'promotion_rule_ids': [(5,)]
+                }
+                vals.append((1, line.id, v))
+            if vals:
+                order.write({'order_line': vals})
+
+    @api.multi
+    def _apply(self, orders):
+        for rule in self:
+            orders = orders.filtered(
+                lambda o, r=rule: r._is_promotion_valid(o))
+            rule._apply_rule_to_order_lines(orders.mapped('order_line'))
+            if rule.rule_type == 'coupon':
+                orders.write({'coupon_promotion_rule_id': rule.id})
+            else:
+                orders.write({'promotion_rule_ids': [(4, rule.id)]})
+
+    @api.multi
+    def _apply_rule_to_order_lines(self, lines):
+        self.ensure_one()
+        lines = lines.filtered(
+            lambda l, r=self: r._is_promotion_valid_for_line(l))
+        if self.promo_type == 'discount':
+            self._apply_discount_to_order_lines(lines)
+        else:
+            raise ValidationError(
+                _('Not supported promotion type %s') % self.promo_type
+            )
+
+    @api.multi
+    def _apply_discount_to_order_lines(self, lines):
+        self.ensure_one()
+        if not self.promo_type == 'discount':
+            return
+        if not self.discount_type == 'percentage':
+            raise ValidationError(
+                _('Discount promotion of type %s is not supported') %
+                self.discount_type)
+
+        lines_by_order = defaultdict(self.env['sale.order.line'].browse)
```

Comment:
browse() ?

Reply:
@rousseldenis arg of defaultdict must be a callable....